File: apps/mobile/ios/App/App/NativeAuth.swift
import AuthenticationServices
import Capacitor
import Foundation
import UIKit

/**
 * The two sign-in sheets a WebView cannot show, and nothing else.
 *
 * This plugin fetches nothing, stores nothing and decides nothing. `signInWithApple` shows
 * Apple's sheet and returns the identity token it produces; `signInWithBrowser` runs a browser
 * outside the app and returns the URL it ended at. Every other part of signing in --- the nonce,
 * the hashing, the token exchange, where the API lives, what to keep --- is in
 * `apps/web/src/nativeAuth.ts`, where the tests are and where `api.ts` already knows the answers.
 *
 * That split is the design rather than an accident of taste. Swift that talks to the API needs a
 * second copy of the origin, the bearer header and the token store, and it is the copy nobody
 * runs in a test. Here, the only thing that cannot be written in TypeScript is the part that is
 * genuinely iOS: a system authorization request, and a browser session out of process.
 *
 * Why out of process at all: Google refuses OAuth in an embedded WebView (`disallowed_useragent`)
 * and `WKAppBoundDomains` refuses to navigate this WebView off our own domains. See docs/IOS.md.
 */
@objc(NativeAuth)
public class NativeAuth: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeAuth"
    public let jsName = "NativeAuth"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "signInWithApple", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "signInWithBrowser", returnType: CAPPluginReturnPromise)
    ]

    /// Held for the life of a request, because `ASAuthorizationController` does not retain its own
    /// delegate and a released one is a sheet that opens and never answers.
    private var appleRequest: AppleRequest?
    /// Same reason: a session that goes out of scope closes itself.
    private var browserSession: ASWebAuthenticationSession?
    /// Where the sheet is anchored. `CAPPlugin` has the bridge's view controller; the window is
    /// what `ASWebAuthenticationSession` asks for.
    private lazy var anchor = PresentationAnchor()

    /**
     * Sign in with Apple, through the system rather than through a browser.
     *
     * `nonceHash` is passed straight to Apple and comes back inside the signed token. It is
     * already a hash when it arrives here --- the server issued the nonce and the web layer
     * hashed it --- so this method has no crypto in it and no opinion about what a nonce is.
     *
     * Only the identity token is returned. The authorization code is of no use to us: the server
     * verifies the token's signature against Apple's public keys and its audience against this
     * app's bundle identifier, and a code would mean a second exchange with nothing to add.
     */
    @objc func signInWithApple(_ call: CAPPluginCall) {
        guard let nonceHash = call.getString("nonceHash"), !nonceHash.isEmpty else {
            call.reject("a nonce is required")
            return
        }

        let request = ASAuthorizationAppleIDProvider().createRequest()
        request.requestedScopes = [.fullName, .email]
        request.nonce = nonceHash

        let controller = ASAuthorizationController(authorizationRequests: [request])
        let pending = AppleRequest(call: call) { [weak self] in
            // Let go once it has answered, so a second sign-in is not held up by the first.
            self?.appleRequest = nil
        }
        appleRequest = pending
        controller.delegate = pending
        controller.presentationContextProvider = anchor
        DispatchQueue.main.async {
            controller.performRequests()
        }
    }

    /**
     * A web handshake in a browser the app does not own.
     *
     * `ASWebAuthenticationSession` is Safari: a real user agent with its own cookie store, which
     * is what makes Google willing to serve it and what puts the navigation outside the reach of
     * app-bound domains. It ends when the browser is sent to a URL whose scheme is `scheme`, and
     * that URL is handed back unread --- what is in it is a question about our own API.
     *
     * Not ephemeral. A shared session means somebody already signed in to Google in Safari is one
     * tap from being signed in here, which is the whole reason to use a real browser.
     */
    @objc func signInWithBrowser(_ call: CAPPluginCall) {
        guard let start = call.getString("url"), let url = URL(string: start) else {
            call.reject("a url is required")
            return
        }
        guard let scheme = call.getString("scheme"), !scheme.isEmpty else {
            call.reject("a callback scheme is required")
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            let session = ASWebAuthenticationSession(
                url: url,
                callbackURLScheme: scheme
            ) { [weak self] callback, error in
                self?.browserSession = nil
                if let callback {
                    call.resolve(["url": callback.absoluteString])
                    return
                }
                // `canceledLogin` is somebody closing the sheet. The word matters: the web layer
                // reads "cancel" out of the message and says nothing to the player about it.
                if let error = error as? ASWebAuthenticationSessionError,
                   error.code == .canceledLogin {
                    call.reject("cancelled")
                    return
                }
                call.reject(error?.localizedDescription ?? "the browser session failed")
            }
            session.presentationContextProvider = self.anchor
            self.browserSession = session
            session.start()
        }
    }
}

/**
 * One Apple authorization request, and the delegate it needs to stay alive for.
 *
 * A class of its own rather than conformances on the plugin, so that two sign-ins in a row cannot
 * share state, and so `done` can drop the reference the moment either callback fires.
 */
private final class AppleRequest: NSObject, ASAuthorizationControllerDelegate {
    private let call: CAPPluginCall
    private let done: () -> Void

    init(call: CAPPluginCall, done: @escaping () -> Void) {
        self.call = call
        self.done = done
    }

    func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithAuthorization authorization: ASAuthorization
    ) {
        defer { done() }
        guard
            let credential = authorization.credential as? ASAuthorizationAppleIDCredential,
            let data = credential.identityToken,
            let token = String(data: data, encoding: .utf8)
        else {
            call.reject("apple returned no identity token")
            return
        }
        call.resolve(["identityToken": token])
    }

    func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithError error: Error
    ) {
        defer { done() }
        // `.canceled` is the common ending, not a fault: somebody looked at the sheet and
        // changed their mind. Reported in a word the web layer recognises.
        if let failure = error as? ASAuthorizationError, failure.code == .canceled {
            call.reject("cancelled")
            return
        }
        call.reject(error.localizedDescription)
    }
}

/**
 * Which window the sheets belong to.
 *
 * `UIApplication.shared.windows` is deprecated and wrong under multiple scenes; the first key
 * window of the first active scene is the app's only window in practice, and is the one the
 * WebView is in.
 */
private final class PresentationAnchor: NSObject, ASWebAuthenticationPresentationContextProviding,
    ASAuthorizationControllerPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        window()
    }

    func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
        window()
    }

    private func window() -> ASPresentationAnchor {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        let found = scenes.flatMap(\.windows).first(where: \.isKeyWindow) ?? scenes.first?.windows.first
        return found ?? ASPresentationAnchor()
    }
}

File: apps/web/src/nativeAuth.ts
import { NATIVE_API_ORIGIN, apiFetch, rememberToken } from './api.js'
import { isNativeApp } from './platform.js'

/**
 * Signing in with Apple or Google from inside the native shell.
 *
 * The web flow cannot work there and `sso.ts` says why at length: a root-relative navigation
 * resolves into the app bundle, app-bound domains refuses to leave, and Google refuses an
 * embedded WebView outright. So the handshake happens **outside** the WebView, in code the
 * operating system runs, and this module is the seam between that and the app.
 *
 * **Everything except the two system sheets is here rather than in Swift.** The plugin fetches
 * nothing and decides nothing: it shows Apple's sheet and returns a token, or runs a browser
 * session and returns the URL it ended at. Each network call goes through `apiFetch`, so the
 * origin, the bearer header and the token store are the same ones the rest of the app uses, and
 * each decision is in TypeScript where the tests are. Swift that talks to an API is Swift that
 * needs its own copy of everything `api.ts` knows.
 *
 * Two doors, because the providers are genuinely different:
 *
 * - **Apple** is `ASAuthorizationAppleIDProvider`: a system sheet, no browser, and a signed
 *   identity token handed straight to the app. We post that token and a nonce the server issued.
 * - **Google** is `ASWebAuthenticationSession`: a real Safari outside the app, running the
 *   ordinary web flow, which ends by sending the browser to `blinkered://auth?code=...`. We trade
 *   that code for a token.
 */

/** What the Swift plugin offers. Two methods, both of them UI and nothing else. */
interface NativeAuthPlugin {
  /**
   * Apple's sheet. `nonceHash` is what goes in the request and comes back inside the token.
   *
   * Rejects when somebody dismisses it, which is not an error worth a message: `cancelled`.
   */
  signInWithApple(options: { nonceHash: string }): Promise<{ identityToken?: unknown }>
  /**
   * A browser session on `url`, ending when it is sent to a URL with `scheme`.
   *
   * Returns that final URL, unparsed. Reading it is this side's job, because what is in it is a
   * decision about our own API rather than about iOS.
   */
  signInWithBrowser(options: { url: string; scheme: string }): Promise<{ url?: unknown }>
}

interface CapacitorWithPlugins {
  readonly Plugins?: { readonly NativeAuth?: Partial<NativeAuthPlugin> }
}

/**
 * The plugin, if this build has it.
 *
 * Checked method by method rather than by asking whether the object exists, because the failure
 * this guards against is real and specific: a shell whose web assets are newer than its Swift.
 * `cap sync` copies the web build into the app, and nothing makes the two halves the same age --
 * an older app with today's bundle would offer the buttons and then call a function that is not
 * there. See docs/IOS.md, "Two commands, every time".
 */
function plugin(): NativeAuthPlugin | null {
  if (!isNativeApp()) return null
  const found = (globalThis as { Capacitor?: CapacitorWithPlugins }).Capacitor?.Plugins?.NativeAuth
  if (typeof found?.signInWithApple !== 'function') return null
  if (typeof found.signInWithBrowser !== 'function') return null
  return found as NativeAuthPlugin
}

/** Whether the shell can offer the provider buttons at all. False in a browser, by design. */
export function nativeSsoAvailable(): boolean {
  return plugin() !== null
}

/**
 * How a native sign-in ended.
 *
 * `reason` is one of the tags the web flow puts in `?signin=`, so `ssoProblem()` already turns
 * every one of them into a sentence in fifty-one languages. `cancelled` is in that vocabulary and
 * is what a dismissed sheet reports, which the dialog shows as nothing at all.
 */
export type NativeSignIn = { readonly ok: true } | { readonly ok: false; readonly reason: string }

const FAILED: NativeSignIn = { ok: false, reason: 'failed' }

async function jsonOf(response: Response): Promise<Record<string, unknown>> {
  return (await response.json().catch(() => ({}))) as Record<string, unknown>
}

/**
 * Keeps the token a sign-in came back with, or says the sign-in did not really happen.
 *
 * A shell holding no token is signed in by nothing -- there is no cookie to fall back on -- so a
 * 200 with no token in it has to be reported rather than believed, the same way `verifyCode`
 * treats the same shape.
 */
async function keep(response: Response): Promise<NativeSignIn> {
  if (!response.ok) {
    const body = await jsonOf(response)
    return { ok: false, reason: typeof body.error === 'string' ? body.error : 'failed' }
  }
  const body = await jsonOf(response)
  if (typeof body.token !== 'string' || body.token === '') return FAILED
  rememberToken(body.token)
  return { ok: true }
}

/** SHA-256 hex, which is what the server expects to find echoed inside Apple's token. */
async function sha256Hex(value: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(value))
  return [...new Uint8Array(digest)].map((byte) => byte.toString(16).padStart(2, '0')).join('')
}

/**
 * Sign in with Apple, natively.
 *
 * The nonce is the server's, hashed before Apple sees it: what travels is not what unlocks. A
 * client-chosen nonce would prove only that the client agrees with itself.
 */
export async function appleNatively(): Promise<NativeSignIn> {
  const native = plugin()
  if (native === null) return FAILED
  try {
    const issued = await jsonOf(await apiFetch('auth/native/nonce', { method: 'POST' }))
    if (typeof issued.nonce !== 'string' || issued.nonce === '') return FAILED

    const credential = await native.signInWithApple({ nonceHash: await sha256Hex(issued.nonce) })
    if (typeof credential.identityToken !== 'string' || credential.identityToken === '') {
      return FAILED
    }

    return await keep(
      await apiFetch('auth/native/apple', {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ identityToken: credential.identityToken, nonce: issued.nonce }),
      }),
    )
  } catch (failure: unknown) {
    // A dismissed sheet arrives here as a rejection, and it is the common case rather than a
    // fault: somebody looked at Apple's sheet and changed their mind.
    return { ok: false, reason: cancelled(failure) ? 'cancelled' : 'failed' }
  }
}

/**
 * Sign in with Google, in a browser outside the app.
 *
 * `?native=1` is what makes the server's callback hand back a code on our own URL scheme instead
 * of setting a cookie on a page nobody will see.
 */
export async function googleNatively(): Promise<NativeSignIn> {
  const native = plugin()
  if (native === null) return FAILED
  try {
    const ended = await native.signInWithBrowser({
      url: startUrlFor('google'),
      scheme: CALLBACK_SCHEME,
    })
    if (typeof ended.url !== 'string') return FAILED

    const returned = new URL(ended.url)
    const error = returned.searchParams.get('error')
    if (error !== null && error !== '') return { ok: false, reason: error }
    const code = returned.searchParams.get('code')
    if (code === null || code === '') return FAILED

    return await keep(
      await apiFetch('auth/native/exchange', {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ code }),
      }),
    )
  } catch (failure: unknown) {
    return { ok: false, reason: cancelled(failure) ? 'cancelled' : 'failed' }
  }
}

/** The scheme the session watches for, which must match `NATIVE_CALLBACK` on the server. */
const CALLBACK_SCHEME = 'blinkered'

/**
 * Where a browser session starts, as an absolute URL.
 *
 * Absolute because this one is handed to iOS rather than to the WebView: a root-relative path
 * would be resolved against `capacitor://localhost` by the very code that is trying to get out of
 * it. This is the one place `sso.ts`'s root-relative `startUrl` cannot be used, and the shape of
 * that mistake is the whole reason this module exists.
 */
function startUrlFor(provider: 'apple' | 'google'): string {
  // `NATIVE_API_ORIGIN` rather than a second constant, so a shell cannot be pointed at two hosts.
  // This is the one request the WebView does not make, because iOS makes it.
  return new URL(`/v1/auth/${provider}?native=1`, NATIVE_API_ORIGIN).toString()
}

/** Whether a rejection is somebody closing a sheet rather than something going wrong. */
function cancelled(failure: unknown): boolean {
  const message = failure instanceof Error ? failure.message : String(failure)
  // Both halves report it in their own words: `ASAuthorizationError.canceled` comes back as
  // "canceled", and `ASWebAuthenticationSessionError.canceledLogin` as "cancelled".
  return /cancel/i.test(message)
}
